{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# determine co-occurrence of interesting variants in genes from PCGC project with Stacia, Casey and Jeeves\n",
    "import gzip\n",
    "import pandas as pd\n",
    "import re\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "\n",
    "\n",
    "### this is the only part of the script you should have to touch regularly ###\n",
    "\n",
    "gene='MED13L'\n",
    "outfile = '../results/med13l_pairs.tsv'\n",
    "gen_vcf_file = 'vcfs/med13l.vcf.gz' # this should be the gzipped VCF for whichever gene from 1000Gs\n",
    "trios_file = '../data/common+rare/cardio277_cases_freq30_list.txt' # this should be the cases (from Stacia)\n",
    "ctrls_file = '../data/rare+rare/cardio277_controls_with_trios.txt' # this should be the controls file (from Stacia)\n",
    "gene_list_file = '../data/rare+rare/cardio277_summary.txt' # this is simply used to get the gene list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# import vcf from 1000 Genomes\n",
    "with gzip.open(gen_vcf_file, 'rb') as f: \n",
    "    gen_df = pd.read_table(f,sep='\\t',comment='#',usecols=[1,*range(9,2513)],index_col=0,header=None)\n",
    "\n",
    "# import trios data\n",
    "trios_df_names = [1,2,3,4]\n",
    "with open(trios_file,'r') as f:\n",
    "    trios_df = pd.read_table(f, sep='\\t', comment='#',names=trios_df_names, index_col=None)\n",
    "    \n",
    "# import gene list\n",
    "with open(gene_list_file,'r') as f:\n",
    "    gen_sum_df = pd.read_table(f, sep='\\t', index_col=None, header=None)\n",
    "gene_list = list(gen_sum_df[0][1:])\n",
    "\n",
    "def get_vars_df(gen_df):\n",
    "    var_pat = re.compile('1\\|1|2\\|2|0\\|1|1\\|0|2\\|0|0\\|2|1\\|2|2\\|1|0\\|3|3\\|0|1\\|3|3\\|1|2\\|3|3\\|2|3\\|3')\n",
    "    non_var_pat = re.compile('0\\|0')\n",
    "\n",
    "    # replace variants with 1 and non-variants with 0\n",
    "\n",
    "    gen_df.replace(to_replace=[var_pat,non_var_pat],value=[1,0],inplace=True,regex=True)\n",
    "\n",
    "    # make sure all values are int\n",
    "\n",
    "    gen_df_int = gen_df.astype(int)\n",
    "    \n",
    "    return(gen_df_int)\n",
    "\n",
    "def get_cooc_df(gen_df_int):\n",
    "\n",
    "    # transpose dataframe to get cooccurrence matrix\n",
    "\n",
    "    coocc = gen_df_int.dot(gen_df_int.T)\n",
    "\n",
    "    # reset diagonal\n",
    "\n",
    "    np.fill_diagonal(coocc.values, 0)\n",
    "\n",
    "    # change index name to avoid later errors\n",
    "\n",
    "    coocc = coocc.reindex(coocc.index.rename('Index'))\n",
    "    \n",
    "    return(coocc)\n",
    "\n",
    "# plot clustermap (if desired)\n",
    "\n",
    "# % matplotlib inline\n",
    "\n",
    "# clusterplot = sns.clustermap(coocc)\n",
    "# sns.heatmap(coocc)\n",
    "\n",
    "# figure out which pairs of variants are present in cases for gene of interest\n",
    "\n",
    "def get_pairs(gene, coocc):\n",
    "    next_gene_index = gene_list.index(gene) + 1\n",
    "    gene_index = trios_df[trios_df[1].str.contains(gene)].index.tolist()\n",
    "    next_gene_index = trios_df[trios_df[1].str.contains(gene_list[next_gene_index])].index.tolist()\n",
    "\n",
    "    rel_df = trios_df.loc[gene_index[0]:next_gene_index[0],:]\n",
    "\n",
    "    # add new column to specify case when 2 SNPs for father or mother\n",
    "    rel_df['check'] = rel_df[1].shift(1)\n",
    "    rel_df['add'] = rel_df[2].shift(-1)\n",
    "    rel_df['check2'] = rel_df[2].shift(2)\n",
    "\n",
    "    father_snps = []\n",
    "    mother_snps = []\n",
    "    for index,row in rel_df.iterrows():\n",
    "        if 'Father' in str(row[1]):\n",
    "            father_snps.append(row[2]) \n",
    "            mother_snps.append(row['add'])\n",
    "        elif len(str(row[1]))==2 and 'Father' in str(row['check']):\n",
    "            father_snps.append(row[2])\n",
    "            mother_snps.append(row['add'])\n",
    "        elif len(str(row[1]))==2 and 'Mother' in str(row['check']):\n",
    "            father_snps.append(row['check2'])\n",
    "            mother_snps.append(row[2])\n",
    "\n",
    "    cooc_freq_list = []\n",
    "    cooc_list = []\n",
    "\n",
    "    counter = -1\n",
    "    for var in father_snps:\n",
    "        counter += 1\n",
    "        mvar = mother_snps[counter]\n",
    "        if int(var) in coocc.index and int(mvar) in coocc.index:\n",
    "            cooc = float(coocc.loc[int(var),int(mvar)])\n",
    "        else:\n",
    "            cooc = np.nan\n",
    "        cooc_list.append(cooc)\n",
    "        cooc_freq_list.append(cooc/2504.0)\n",
    "    cooc_of_interest = pd.DataFrame({'Father SNP':father_snps,'Mother SNP':mother_snps,'Cooccurrence freq in 1000 Genomes':cooc_freq_list,'Cooc in 1000 Genomes':cooc_list})\n",
    "    cooc_of_interest = cooc_of_interest.sort_values(by='Cooccurrence freq in 1000 Genomes', ascending=False)\n",
    "    return(cooc_of_interest)\n",
    "\n",
    "def save_cooc_results(gene, outfile):\n",
    "    gen_df_int = get_vars_df(gen_df)\n",
    "    coocc_df = get_cooc_df(gen_df_int)\n",
    "    pairs_df = get_pairs(gene, coocc_df)\n",
    "    pairs_df.to_csv(outfile,sep='\\t')\n",
    "    return(pairs_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "out_df = save_cooc_results(gene, outfile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# check for these SNPs in controls\n",
    "\n",
    "ctrl_names = [1,2,3,4]\n",
    "with open(ctrls_file, 'r') as f:\n",
    "    ctrl_df = pd.read_table(ctrls_file, names=ctrl_names, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# # how often do these SNPs occur individually in the controls?\n",
    "\n",
    "# next_gene_index = gene_list.index(gene) + 1\n",
    "# gene_index = ctrl_df[ctrl_df[1].str.contains(gene)].index.tolist()\n",
    "# next_gene_index = ctrl_df[ctrl_df[1].str.contains(gene_list[next_gene_index])].index.tolist()\n",
    "\n",
    "# rel_ctrls = ctrl_df.loc[gene_index[0]:next_gene_index[0],:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     Cooc in 1000 Genomes  Cooccurrence freq in 1000 Genomes Father SNP  \\\n",
      "52                   84.0                           0.033546   23859610   \n",
      "16                   84.0                           0.033546   23859610   \n",
      "82                   84.0                           0.033546   23859610   \n",
      "74                   84.0                           0.033546   23859610   \n",
      "70                   84.0                           0.033546   23859610   \n",
      "64                   84.0                           0.033546   23859610   \n",
      "62                   84.0                           0.033546   23874507   \n",
      "46                   84.0                           0.033546   23859610   \n",
      "44                   84.0                           0.033546   23859610   \n",
      "39                   84.0                           0.033546   23874507   \n",
      "36                   84.0                           0.033546   23874507   \n",
      "33                   84.0                           0.033546   23859610   \n",
      "30                   84.0                           0.033546   23874507   \n",
      "100                  84.0                           0.033546   23874507   \n",
      "11                   43.0                           0.017173   23859610   \n",
      "28                   43.0                           0.017173   23876267   \n",
      "1                    43.0                           0.017173   23876267   \n",
      "89                   43.0                           0.017173   23876267   \n",
      "13                   43.0                           0.017173   23859610   \n",
      "10                   43.0                           0.017173   23859610   \n",
      "4                    43.0                           0.017173   23876267   \n",
      "48                   43.0                           0.017173   23876267   \n",
      "79                   43.0                           0.017173   23876267   \n",
      "81                   43.0                           0.017173   23859610   \n",
      "98                   43.0                           0.017173   23859610   \n",
      "19                   43.0                           0.017173   23859610   \n",
      "76                   31.0                           0.012380   23876267   \n",
      "67                   31.0                           0.012380   23876267   \n",
      "47                   31.0                           0.012380   23874507   \n",
      "63                   31.0                           0.012380   23874507   \n",
      "..                    ...                                ...        ...   \n",
      "15                    0.0                           0.000000   23855715   \n",
      "88                    0.0                           0.000000   23874507   \n",
      "78                    0.0                           0.000000   23876267   \n",
      "72                    0.0                           0.000000   23866269   \n",
      "60                    0.0                           0.000000   23868065   \n",
      "49                    0.0                           0.000000   23874507   \n",
      "41                    0.0                           0.000000   23871683   \n",
      "40                    0.0                           0.000000   23871683   \n",
      "32                    0.0                           0.000000   23874507   \n",
      "20                    0.0                           0.000000   23876267   \n",
      "17                    0.0                           0.000000   23874507   \n",
      "59                    0.0                           0.000000   23868065   \n",
      "14                    NaN                                NaN   23858266   \n",
      "18                    NaN                                NaN   23876267   \n",
      "22                    NaN                                NaN   23874507   \n",
      "23                    NaN                                NaN   23855714   \n",
      "25                    NaN                                NaN   23876267   \n",
      "29                    NaN                                NaN   23876267   \n",
      "31                    NaN                                NaN   23855309   \n",
      "35                    NaN                                NaN   23876318   \n",
      "38                    NaN                                NaN   23874507   \n",
      "50                    NaN                                NaN   23876267   \n",
      "51                    NaN                                NaN   23868075   \n",
      "66                    NaN                                NaN   23874507   \n",
      "73                    NaN                                NaN   23858107   \n",
      "75                    NaN                                NaN   23874507   \n",
      "77                    NaN                                NaN   23859429   \n",
      "84                    NaN                                NaN   23873509   \n",
      "85                    NaN                                NaN   23859610   \n",
      "87                    NaN                                NaN   23862646   \n",
      "\n",
      "    Mother SNP  \n",
      "52    23874507  \n",
      "16    23874507  \n",
      "82    23874507  \n",
      "74    23874507  \n",
      "70    23874507  \n",
      "64    23874507  \n",
      "62    23859610  \n",
      "46    23874507  \n",
      "44    23874507  \n",
      "39    23859610  \n",
      "36    23859610  \n",
      "33    23874507  \n",
      "30    23859610  \n",
      "100   23859610  \n",
      "11    23876267  \n",
      "28    23859610  \n",
      "1     23859610  \n",
      "89    23859610  \n",
      "13    23876267  \n",
      "10    23876267  \n",
      "4     23859610  \n",
      "48    23859610  \n",
      "79    23859610  \n",
      "81    23876267  \n",
      "98    23876267  \n",
      "19    23876267  \n",
      "76    23874507  \n",
      "67    23874507  \n",
      "47    23876267  \n",
      "63    23876267  \n",
      "..         ...  \n",
      "15    23874507  \n",
      "88    23873940  \n",
      "78    23863383  \n",
      "72    23874507  \n",
      "60    23858697  \n",
      "49    23873940  \n",
      "41    23866269  \n",
      "40    23862997  \n",
      "32    23873940  \n",
      "20    23873561  \n",
      "17    23868065  \n",
      "59    23855645  \n",
      "14    23859610  \n",
      "18    23851737  \n",
      "22    23874948  \n",
      "23    23874507  \n",
      "25    23855609  \n",
      "29    23874853  \n",
      "31    23876267  \n",
      "35    23874507  \n",
      "38    23851639  \n",
      "50    23872624  \n",
      "51    23869511  \n",
      "66    23862169  \n",
      "73    23874507  \n",
      "75    23855165  \n",
      "77    23874507  \n",
      "84    23876267  \n",
      "85    23871676  \n",
      "87    23872624  \n",
      "\n",
      "[101 rows x 4 columns]\n"
     ]
    }
   ],
   "source": [
    "print(out_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python [python3]",
   "language": "python",
   "name": "Python [python3]"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
